libyul/optimiser/RedundantStoreEliminator.cpp

```
@@ -0,0 +1,452 @@
+{
+	RedundantStoreBase::operator()(_functionCall);
+
+	for (Operation const& op: operationsFromFunctionCall(_functionCall))
+		applyOperation(op);
+
+	ControlFlowSideEffects sideEffects;
+	if (auto builtin = m_dialect.builtin(_functionCall.functionName.name))
+		sideEffects = builtin->controlFlowSideEffects;
+	else
+		sideEffects = m_controlFlowSideEffects.at(_functionCall.functionName.name);
+
+	if (!sideEffects.canContinue)
+	{
+		changeUndecidedTo(State::Unused, Location::Memory);
+
+		changeUndecidedTo(
+			sideEffects.canTerminate ?
+			State::Used :
+			State::Unused,
+			Location::Storage
+		);
+	}
+}
+
+void RedundantStoreEliminator::operator()(FunctionDefinition const& _functionDefinition)
+{
+	ScopedSaveAndRestore storeOperations(m_storeOperations, {});
+	RedundantStoreBase::operator()(_functionDefinition);
+}
+
+
+void RedundantStoreEliminator::operator()(Leave const&)
+{
+	changeUndecidedTo(State::Used);
+}
+
+void RedundantStoreEliminator::visit(Statement const& _statement)
+{
+	using evmasm::Instruction;
+
+	RedundantStoreBase::visit(_statement);
+
+	auto const* exprStatement = get_if<ExpressionStatement>(&_statement);
+	if (!exprStatement)
+		return;
+
+	FunctionCall const* funCall = get_if<FunctionCall>(&exprStatement->expression);
+	if (!funCall)
+		return;
+	optional<Instruction> instruction = toEVMInstruction(m_dialect, funCall->functionName.name);
+	if (!instruction)
+		return;
+
+	if (!ranges::all_of(funCall->arguments, [](Expression const& _expr) -> bool {
+		return get_if<Identifier>(&_expr) || get_if<Literal>(&_expr);
+	}))
+		return;
+
+	if (
+		*instruction == Instruction::SSTORE ||
+		(!m_ignoreMemory && (
+			*instruction == Instruction::EXTCODECOPY ||
+			*instruction == Instruction::CODECOPY ||
+			*instruction == Instruction::CALLDATACOPY ||
+			*instruction == Instruction::RETURNDATACOPY ||
+			*instruction == Instruction::MSTORE ||
+			*instruction == Instruction::MSTORE8
+		))
+	)
+	{
```

Wouldn't it be better to put it in `operator()(FunctionCall)` instead? You would get rid of all those checks above this way.

It would also be a bit more robust against future code modifications. I think it's here because none of those instructions returns value and it makes sense as long as that's the case but it's also a bit dangerous in that you have to be aware of this assumption when modifying this list - if you add an instruction that can actually be used outside of `ExpressionStatement`s, any such use will be silently ignored.

---

It has to be here because in the end, we have to store pointers to statements and not pointers to function calls.

---

Oh, ok. I guess we could change the base to store pointers to function calls but it's probably not worth going that deep for so little gain.